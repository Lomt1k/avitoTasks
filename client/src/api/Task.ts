import { z } from "zod";
import { AssigneeSchema } from "./User";
import { api, validateResponse } from "./api";

export enum TaskStatus {
  Backlog = 'Backlog',
  InProgress = 'InProgress',
  Done = 'Done',
};

export enum TaskPriority {
  Low = 'Low',
  Medium = 'Medium',
  High = 'High',
}

export const TaskSchema = z.object({
  assignee: AssigneeSchema,
  boardId: z.number().optional(),
  boardName: z.string().optional(),
  description: z.string(),
  id: z.number(),
  priority: z.enum(Object.values(TaskPriority) as [string, ...string[]]),
  status: z.enum(Object.values(TaskStatus) as [string, ...string[]]),
  title: z.string(),
});

export const TasksResponseSchema = z.object({
  data: z.array(TaskSchema),
});

export const CreateTaskSchema = z.object({
  assigneeId: z.number(),
  boardId: z.number(),
  title: z.string()
    .nonempty('Необходимо указать название')
    .max(100, 'Должно быть не более 100 символов'),
  description: z.string()
    .nonempty('Необходимо указать описание')
    .max(500, 'Должно быть не более 500 символов'),
  priority: z.enum(Object.values(TaskPriority) as [string, ...string[]]),
});

export const UpdateTaskSchema = z.object({
  assigneeId: z.number(),
  title: z.string()
    .nonempty('Необходимо указать название')
    .max(100, 'Должно быть не более 100 символов'),
    description: z.string()
    .nonempty('Необходимо указать описание')
    .max(500, 'Должно быть не более 500 символов'),
  priority: z.enum(Object.values(TaskPriority) as [string, ...string[]]),
  status: z.enum(Object.values(TaskStatus) as [string, ...string[]]),
});

export type Task = z.infer<typeof TaskSchema>;

export type CreateTaskData = z.infer<typeof CreateTaskSchema>;

export type UpdateTaskData = z.infer<typeof UpdateTaskSchema>;

export const fetchTasksFromBoard = async (boardId: number): Promise<Task[]> => {
  const response = await api.get(`/boards/${boardId}`);
  validateResponse(response);
  const result = TasksResponseSchema.safeParse(response.data);
  if (!result.data) {
    console.error(result.error);
    throw result.error;
  }
  return result.data.data;
}

export const fetchAllTasks = async (): Promise<Task[]> => {
  const response = await api.get(`/tasks`);
  validateResponse(response);
  const result = TasksResponseSchema.safeParse(response.data);
  if (!result.data) {
    console.error(result.error);
    throw result.error;
  }
  return result.data.data;
}

export const fetchCreateTask = async (data: CreateTaskData): Promise<void> => {
  const response = await api.post('/tasks/create', data);
  validateResponse(response);
}

export const fetchUpdateTask = async (taskId: number, data: UpdateTaskData): Promise<void> => {
  const response = await api.put(`/tasks/update/${taskId}`, data);
  validateResponse(response);
}