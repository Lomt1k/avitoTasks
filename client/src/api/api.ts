import axios, { AxiosResponse } from "axios";
import { z } from "zod";

export const api = axios.create({
  baseURL: 'http://localhost:8080/api/v1',
  timeout: 1000,
  validateStatus: () => true,
});

export const ErrorMessageSchema = z.object({
  error: z.string(),
  message: z.string(),
});

export type ErrorMessage = z.infer<typeof ErrorMessageSchema>;

export const validateResponse = (response: AxiosResponse): void => {
  if (response.status === 200) {
    return;
  }

  const result = ErrorMessageSchema.safeParse(response.data);
  if (result.data) {
    const error = new Error(result.data.message);
    error.name = result.data.error;
    console.error(error);
    throw error;
  }

  const error = new Error(`${response.status} ${response.statusText}`);
  error.name = 'Network Error';
  console.error(error);
  throw error;
}