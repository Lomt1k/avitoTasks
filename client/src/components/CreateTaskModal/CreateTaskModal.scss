.create-task-modal {
  min-width: 400px;

  &__heading {
    margin: 0 0 40px;
    font-size: 20px;
    font-weight: 600;
  }
}