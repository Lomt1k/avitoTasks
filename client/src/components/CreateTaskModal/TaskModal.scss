.task-modal {
  width: 100%;
  max-width: 700px;

  &__heading {
    margin: 0 0 30px;
    font-size: 20px;
    font-weight: 600;
  }

  &__fields {
    max-width: 400px;
    display: flex;
    flex-direction: column;
    gap: 20px;
    margin: 0 0 30px;
  }

  &__buttons {
    display: flex;
    gap: 20px;
  }

  &__submit-btn {
    margin-left: auto;
  }
}