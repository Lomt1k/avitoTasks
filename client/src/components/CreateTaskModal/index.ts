export { CreateTaskModal } from './CreateTaskModal';
export { EditTaskModal } from './EditTaskModal';