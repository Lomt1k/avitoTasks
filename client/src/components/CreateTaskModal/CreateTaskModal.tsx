import { FC } from 'react';
import { Modal } from '../ui';
import { Task } from '../../api/Task';
import './CreateTaskModal.scss';

type CreateTaskModalProps = {
  onClickClose: () => void;
  task?: Task;
}

const CreateTaskModal: FC<CreateTaskModalProps> = ({ onClickClose, task }) => {
  return (
    <Modal className='create-task-modal' onClickClose={onClickClose}>
      <h3 className="create-task-modal__heading">
        {task ? 'Редактирование задачи' : 'Создание задачи'}
      </h3>
    </Modal>
  )
}

export { CreateTaskModal };