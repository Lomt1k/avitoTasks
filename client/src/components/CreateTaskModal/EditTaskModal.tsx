import { FC } from 'react';
import { Button, ButtonLink, Input, Modal, Select, SelectItem, TextArea } from '../ui';
import { Task, TaskPriority, TaskStatus } from '../../api/Task';
import { getPriorityLocalization, getStatusLocalization } from '../../utils/TaskHelper';
import { observer } from 'mobx-react-lite';
import { useEditTaskForm } from '../../hooks';
import './TaskModal.scss';

type EditTaskModalProps = {
  onClickClose: () => void;
  task: Task;
}

const priorities = Object.values(TaskPriority);
const statuses = Object.values(TaskStatus);

const EditTaskModal: FC<EditTaskModalProps> = observer(({ onClickClose, task }) => {
  const { register, onSubmit, errors, users, isUsersError, boardName } = useEditTaskForm(task, onClickClose);

  return (
    <Modal className='task-modal' onClickClose={onClickClose}>
      <form className='task-modal__form' onSubmit={onSubmit}>
        <h3 className="task-modal__heading">Редактирование задачи</h3>
        <div className="task-modal__fields">
          <Input
            placeholder='Название'
            type='text'
            error={errors.title?.message}
            {...register('title')}
          />
          <TextArea
            placeholder='Описание'
            error={errors.description?.message}
            {...register('description')}
          />
          {boardName &&
            <Select disabled>
              <SelectItem value={boardName}>
                {boardName}
              </SelectItem>
            </Select>
          }
          <Select
            error={errors.priority?.message}
            {...register('priority')}
          >
            {priorities.map(priority => (
              <SelectItem key={priority} value={priority.toString()}>
                {getPriorityLocalization(priority)}
              </SelectItem>
            ))}
          </Select>
          <Select
            error={errors.status?.message}
            {...register('status')}
          >
            {statuses.map(status => (
              <SelectItem key={status} value={status.toString()}>
                {getStatusLocalization(status)}
              </SelectItem>
            ))}
          </Select>
          <Select
            error={isUsersError ? 'Произошла ошибка при подгрузке пользователей' : errors.assigneeId?.message}
            {...register('assigneeId', { setValueAs: value => Number(value) })}
          >
            {users && users.map(user => (
              <SelectItem key={user.id} value={user.id.toString()}>
                {user.fullName}
              </SelectItem>
            ))}
          </Select>
        </div>
        <div className="task-modal__buttons">
          {task.boardId &&
            <ButtonLink secondary to={`/board/${task.boardId}`}>
              Перейти на доску
            </ButtonLink>
          }
          <Button className='task-modal__submit-btn' submit>Обновить</Button>
        </div>
      </form>
    </Modal>
  )
})

export { EditTaskModal };