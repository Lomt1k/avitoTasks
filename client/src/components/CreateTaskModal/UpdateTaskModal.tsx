import { FC, useEffect } from 'react';
import { Button, ButtonLink, Input, Modal, Select, SelectItem, TextArea } from '../ui';
import { fetchUpdateTask, Task, TaskPriority, TaskStatus, UpdateTaskData, UpdateTaskSchema } from '../../api/Task';
import { zodResolver } from '@hookform/resolvers/zod';
import { useForm } from 'react-hook-form';
import { getPriorityLocalization, getStatusLocalization } from '../../utils/TaskHelper';
import { observer } from 'mobx-react-lite';
import RootStore from '../../store/RootStore';
import { useMutation, useQueryClient } from '@tanstack/react-query';
import { useUsers } from '../../hooks';
import './TaskModal.scss';

type UpdateTaskModalProps = {
  onClickClose: () => void;
  task: Task;
}

const priorities = Object.values(TaskPriority);
const statuses = Object.values(TaskStatus);

const UpdateTaskModal: FC<UpdateTaskModalProps> = observer(({ onClickClose, task }) => {
  const { register, handleSubmit, setValue, formState: { errors } } = useForm<UpdateTaskData>({
    resolver: zodResolver(UpdateTaskSchema)
  });
  const { data: users, isError: isUsersError } = useUsers();
  const queryClient = useQueryClient();
  const boardName = task.boardName ?? RootStore.boards.currentBoard?.name;

  const updateTaskMutation = useMutation({
    mutationKey: ['task', 'update', task.id],
    mutationFn: ({ id, taskUpdate }: { id: number; taskUpdate: UpdateTaskData }) => {
      return fetchUpdateTask(id, taskUpdate);
    },
    onSuccess() {
      queryClient.invalidateQueries({ queryKey: ['tasks'] });
      onClickClose();
    }
  });

  useEffect(() => {
    if (users && task.assignee.id) {
      setValue('assigneeId', task.assignee.id);
    }
  }, [users, task.assignee.id]);

  return (
    <Modal className='task-modal' onClickClose={onClickClose}>
      <form onSubmit={handleSubmit(taskUpdate => {
        updateTaskMutation.mutate({ id: task.id, taskUpdate });
      })}
      >
        <h3 className="task-modal__heading">Редактирование задачи</h3>
        <div className="task-modal__fields">
          <Input
            placeholder='Название'
            type='text'
            defaultValue={task.title}
            error={errors.title?.message}
            {...register('title')}
          />
          <TextArea
            placeholder='Описание'
            defaultValue={task.description}
            error={errors.description?.message}
            {...register('description')}
          />
          {boardName &&
            <Select disabled>
              <SelectItem value={boardName}>
                {boardName}
              </SelectItem>
            </Select>
          }
          <Select
            defaultValue={task.priority}
            error={errors.priority?.message}
            {...register('priority')}
          >
            {priorities.map(priority => (
              <SelectItem key={priority} value={priority.toString()}>
                {getPriorityLocalization(priority)}
              </SelectItem>
            ))}
          </Select>
          <Select
            defaultValue={task.status}
            error={errors.status?.message}
            {...register('status')}
          >
            {statuses.map(status => (
              <SelectItem key={status} value={status.toString()}>
                {getStatusLocalization(status)}
              </SelectItem>
            ))}
          </Select>
          <Select
            defaultValue={task.assignee.id}
            error={isUsersError ? 'Произошла ошибка при подгрузке пользователей' : errors.assigneeId?.message}
            {...register('assigneeId', { setValueAs: value => Number(value) })}
          >
            {users && users.map(user => (
              <SelectItem key={user.id} value={user.id.toString()} defaultChecked={user.id === task.assignee.id}>
                {user.fullName}
              </SelectItem>
            ))}
          </Select>
        </div>
        <div className="task-modal__buttons">
          {task.boardId &&
            <ButtonLink secondary to={`/board/${task.boardId}`}>
              Перейти на доску
            </ButtonLink>
          }
          <Button className='task-modal__submit-btn' submit>
            Обновить
          </Button>
        </div>
      </form>
    </Modal>
  )
})

export { UpdateTaskModal };