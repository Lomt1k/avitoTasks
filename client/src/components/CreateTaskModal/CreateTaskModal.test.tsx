import { describe, it, expect, vi } from 'vitest';
import { render } from '@testing-library/react';
import { CreateTaskModal } from './CreateTaskModal';
import { getPriorityLocalization } from '../../utils/TaskHelper';
import { TaskPriority } from '../../api/Task';

const onClickClose = vi.fn();
const priorities = Object.values(TaskPriority);

vi.mock('../../hooks', () => ({
  useCreateTaskForm: vi.fn().mockReturnValue({
    register: vi.fn(),
    onSubmit: vi.fn(),
    errors: {},
    users: [
      { id: 1, fullName: 'User 1' },
      { id: 2, fullName: 'User 2' }
    ],
    isUsersError: false,
    boards: [
      { id: 1, name: 'Board 1' },
      { id: 2, name: 'Board 2' }
    ]
  }),
}));

describe('CreateTaskModal', () => {
  it('Рендерится корректно', () => {
    const root = document.createElement('div');
    root.id = 'root';
    document.body.append(root);
    const { getByText, getByPlaceholderText } = render(<CreateTaskModal onClickClose={onClickClose} />);

    expect(getByText('Создание задачи')).toBeInTheDocument();
    expect(getByPlaceholderText('Название')).toBeInTheDocument();
    expect(getByPlaceholderText('Описание')).toBeInTheDocument();
    expect(getByText('Создать')).toBeInTheDocument();
    expect(getByText('Board 1')).toBeInTheDocument();
    expect(getByText('Board 2')).toBeInTheDocument();
    expect(getByText(getPriorityLocalization(priorities[0]))).toBeInTheDocument();
    expect(getByText(getPriorityLocalization(priorities[1]))).toBeInTheDocument();
    expect(getByText('User 1')).toBeInTheDocument();
    expect(getByText('User 2')).toBeInTheDocument();
  });

  it('Снапшот', () => {
    const root = document.createElement('div');
    root.id = 'root';
    document.body.append(root);
    const { container } = render(<CreateTaskModal onClickClose={onClickClose} />);
    expect(container).toMatchSnapshot();
  });
});
