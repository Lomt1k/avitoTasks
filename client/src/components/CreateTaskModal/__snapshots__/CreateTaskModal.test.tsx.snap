// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreateTaskModal > Снапшот 1`] = `<div />`;
